File: algorithms/Dijkstra.py
from queue import PriorityQueue
import networkx as nx
from typing import Any, Optional, List, Tuple
import math
import sys
import os

# Add parent directory to path for imports
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from vis.tracer import get_tracer

def dijkstra(
    G: nx.Graph, 
    source: Any, 
    target: Any
) -> Optional[Tuple[List[Any], float, List[Any]]]:
    """
    Dijkstra's algorithm for finding shortest paths.
    
    Parameters
    ----------
    G : nx.Graph
        Graph with optional 'weight' edge attribute.
    source : Any
        Start node.
    target : Any
        Goal node.
    
    Returns
    -------
    (path, total_weight, visit_order) or None if no path exists.
    """
    if source not in G or target not in G:
        return None
    
    tr = get_tracer()
    pq = PriorityQueue()
    dist = {node: float('inf') for node in G.nodes}
    parent = {source: None}
    visit_order = []
    
    pq.put((0, source))
    dist[source] = 0

    while not pq.empty():
        du, u = pq.get()
        
        if du > dist[u]:
            continue
        
        visit_order.append(u)
        
        # Emit expand event
        tr.emit("Dijkstra", "expand", {
            "current": u,
            "cost": du,
            "parent_map": dict(parent),
        })
        
        if u == target:
            # Reconstruct path
            path = []
            cur = target
            while cur is not None:
                path.append(cur)
                cur = parent.get(cur)
            path.reverse()
            
            tr.emit("Dijkstra", "goal", {
                "path": path,
                "cost": dist[target],
                "parent_map": dict(parent),
            })
            
            return path, dist[target], visit_order

        for v in G.neighbors(u):
            w = G[u][v].get('weight', 1)
            alt = dist[u] + w
            
            if alt < dist[v]:
                old_dist = dist[v]
                dist[v] = alt
                parent[v] = u
                pq.put((dist[v], v))
                
                tr.emit("Dijkstra", "relax", {
                    "from": u,
                    "to": v,
                    "weight": w,
                    "old_dist": old_dist if old_dist != float('inf') else None,
                    "new_dist": alt,
                    "parent_map": dict(parent),
                })
    
    tr.emit("Dijkstra", "no_path", {
        "source": source,
        "target": target,
    })
    
    return None


if __name__ == "__main__":
    G = nx.Graph()
    G.add_weighted_edges_from([
        ('A', 'B', 4),
        ('A', 'C', 2),
        ('B', 'C', 1),
        ('B', 'D', 5),
        ('C', 'D', 8),
        ('C', 'E', 10),
        ('D', 'E', 2),
        ('D', 'Z', 6),
        ('E', 'Z', 3)
    ])

    result = dijkstra(G, 'A', 'Z')

    if result is None:
        print("Không có đường đi từ A đến Z")
    else:
        path, total_weight, visit_order = result
        print("Đường đi ngắn nhất A → Z:", path)
        print("Tổng trọng số:", total_weight)
        print("Thứ tự thăm:", visit_order)

File: webapp/src/algorithms/dijkstra.js
// Dijkstra's algorithm with step-by-step visualization

import { getNeighbors } from '../utils/graphUtils';

/**
 * Min-heap priority queue implementation
 */
class PriorityQueue {
  constructor() {
    this.heap = [];
  }

  push(item, priority) {
    this.heap.push({ item, priority });
    this.bubbleUp(this.heap.length - 1);
  }

  pop() {
    if (this.heap.length === 0) return null;
    const top = this.heap[0];
    const last = this.heap.pop();
    if (this.heap.length > 0) {
      this.heap[0] = last;
      this.bubbleDown(0);
    }
    return top;
  }

  isEmpty() {
    return this.heap.length === 0;
  }

  bubbleUp(index) {
    while (index > 0) {
      const parent = Math.floor((index - 1) / 2);
      if (this.heap[parent].priority <= this.heap[index].priority) break;
      [this.heap[parent], this.heap[index]] = [this.heap[index], this.heap[parent]];
      index = parent;
    }
  }

  bubbleDown(index) {
    while (true) {
      const left = 2 * index + 1;
      const right = 2 * index + 2;
      let smallest = index;
      if (left < this.heap.length && this.heap[left].priority < this.heap[smallest].priority) {
        smallest = left;
      }
      if (right < this.heap.length && this.heap[right].priority < this.heap[smallest].priority) {
        smallest = right;
      }
      if (smallest === index) break;
      [this.heap[smallest], this.heap[index]] = [this.heap[index], this.heap[smallest]];
      index = smallest;
    }
  }

  toArray() {
    return this.heap.map(h => h.item);
  }
}

/**
 * Dijkstra's algorithm implementation with event emission for visualization
 * @param {Object} graph - Graph object with nodes and edges
 * @param {string} source - Source node id
 * @param {string} target - Target node id
 * @returns {Object} - Result with path, cost, steps, and visited order
 */
export function dijkstra(graph, source, target) {
  const steps = [];
  const dist = new Map();
  const parent = new Map();
  const visited = new Set();
  const pq = new PriorityQueue();
  const visitOrder = [];

  // Initialize distances
  for (const [id] of graph.nodes) {
    dist.set(id, Infinity);
  }
  dist.set(source, 0);
  parent.set(source, null);
  pq.push(source, 0);

  // Initial step
  steps.push({
    type: 'init',
    current: null,
    distances: Object.fromEntries(dist),
    visited: [...visited],
    frontier: pq.toArray(),
    parent: Object.fromEntries(parent),
    message: `Starting Dijkstra from node ${source}`,
  });

  while (!pq.isEmpty()) {
    const { item: current, priority: currentDist } = pq.pop();

    if (visited.has(current)) continue;
    if (currentDist > dist.get(current)) continue;

    visited.add(current);
    visitOrder.push(current);

    // Expand step
    steps.push({
      type: 'expand',
      current,
      currentDistance: dist.get(current),
      distances: Object.fromEntries(dist),
      visited: [...visited],
      frontier: pq.toArray(),
      parent: Object.fromEntries(parent),
      message: `Expanding node ${current} (distance: ${dist.get(current)})`,
    });

    // Check if we reached the target
    if (current === target) {
      const path = reconstructPath(parent, target);
      const cost = dist.get(target);

      steps.push({
        type: 'goal',
        current,
        path,
        cost,
        distances: Object.fromEntries(dist),
        visited: [...visited],
        message: `Found target ${target}! Shortest path cost: ${cost}`,
      });

      return {
        found: true,
        path,
        cost,
        steps,
        visitOrder,
        nodesVisited: visited.size,
      };
    }

    // Relax edges
    const neighbors = getNeighbors(graph, current);
    for (const { node: neighbor, weight } of neighbors) {
      if (visited.has(neighbor)) continue;

      const newDist = dist.get(current) + weight;

      if (newDist < dist.get(neighbor)) {
        const oldDist = dist.get(neighbor);
        dist.set(neighbor, newDist);
        parent.set(neighbor, current);
        pq.push(neighbor, newDist);

        steps.push({
          type: 'relax',
          current,
          neighbor,
          oldDistance: oldDist,
          newDistance: newDist,
          weight,
          distances: Object.fromEntries(dist),
          visited: [...visited],
          frontier: pq.toArray(),
          parent: Object.fromEntries(parent),
          message: `Relaxing edge ${current} → ${neighbor}: ${oldDist === Infinity ? '∞' : oldDist} → ${newDist}`,
        });
      }
    }
  }

  // No path found
  steps.push({
    type: 'no_path',
    visited: [...visited],
    distances: Object.fromEntries(dist),
    message: `No path found from ${source} to ${target}`,
  });

  return {
    found: false,
    path: [],
    cost: Infinity,
    steps,
    visitOrder,
    nodesVisited: visited.size,
  };
}

/**
 * Reconstruct path from parent map
 */
function reconstructPath(parent, target) {
  const path = [];
  let current = target;
  while (current !== null) {
    path.unshift(current);
    current = parent.get(current);
  }
  return path;
}

export default dijkstra;
